### PR TITLE
+Add cuberoot function

### PR DESCRIPTION
**Pull request by Hallberg-NOAA**
Added the new functions `cuberoot()` and `intrinsic_functions_unit_tests()` to the MOM_intrinsic_functions module, and call intrinsic_functions_unit_tests from unit_tests to confirm that this new function works as intended.  Separately, cuberoot was tested by replacing expressions like `A**(1./3.)` with `cuberoot(A)` in MOM_energetic_PBL and verifying that the answers only change at roundoff, but that it can give bitwise identical results when the argument is scaled by an integer power of 8 and then unscaled by the corresponding integer power of 2, but that change will occur in a subsequent commit as it can change answers depending on an ANSWER_DATE flag.  With this commit, cuberoot is not yet being used so all answers are bitwise identical, although there are new publicly visible routines.

### CODE REVIEW

**Review by marshallward — CHANGES_REQUESTED** [review]
I am seeing a small issue with the `cuberoot` function.  If I evaluate `cuberoot(0.125)` then it gives me `1.0` rather than `0.5`.  Any other value that is a power of 0.125 (e.g. `2.**-6`) shows the same problem.

I believe it is due to the `elseif` in this if-block:
```
    if (asx > 1.0) then
      asx = 0.125 * asx ; ex_3 = ex_3 + 1
    elseif (asx <= 0.125) then
      asx = 8.0 * asx
    endif
```
Values greater that 1.0 correct their exponent (`ex_3 += 1`) after scaling down, but values below 0.125 do not include this index correct after scaling up.  It looks like the bug can be fixed with `ex_3 = ex_3 - 1`.

----

I looked at the behavior of the uncorrected `asx` function, `scale(abs(x), -3 * (exponent(x) / 3))`, and it appears to already be bounded between 0.125 and 4; only powers of 0.125 are evaluated to 0.125, so it's an extremely small number of values that would be affected by this missing term.

I considered removing the entire elseif block, but it seemed to introduce a minor roundoff that could not be removed by iteration.  I don't know how important it is to reproduce the answer to machine precision, but something else to consider.  (Are all values correct to machine precision?)

**Comment by Hallberg-NOAA** [comment]
Thank you for pointing out the bug with the `(asx <= 0.125)` case.  I just fixed this in a force-push.

The Newton's method iterations do give solutions that are accurate to machine precision, but in some cases when we have arrived at the answer the last bit bounces back and forth around the exact (irrational) solution with added iterations. It is worth noting that the Newton's method convergence from above is monotonic (at least mathematically). If the first guess is an under-estimate, Newton's method will overshoot (a little) and then all subsequent iterations converge monotonically from above.  So the case where the last bit is jumping around, it means that we are in a perpetual roundoff-induced two-step limit cycle.

 Given this limiting behavior at effective convergence, we could choose to break the iterations when they are at the minimum of two successive solutions that are within roundoff of each other, but it would involve adding lines to store the previous value, and then to compare the solutions to make sure that they are not increasing, and this would add to the cost of the function without giving any meaningful increase in accuracy. Is it worth the (hopefully small) extra cost to avoid incorporating any arbitrary choices about the number of iterations to apply, even if it means that our version of the cube root would then always become an under-estimate of some irrational solutions?

**Comment by ashao** [comment]
What about checking the size of the Newton step as a stopping criterion if you're ok with sacrificing a little bit of extra accuracy in the solution? Something like:
```
if (ABS(update) < 2.0*10.0*EPSILON(root)) exit
```
(factor of 2.0 for the roundoff calculation and a factor of 10.0 to go one order of magnitude above roundoff).

**Comment by marshallward** [comment]
I have a suggestion to keep the value of `asx` bounded between 0.125 and 1.  The current form has some odd behavior due to Fortran-isms related to integer division.

The current formula, `asx = scale(abs(x), -3*(exponent(x) / 3))`, looks something like this:

old_asx

After which the if-blocks are required to bound the values above 1 to something within 0.125 and 1.  (I did not look at the actual curve, but surely it will be some piecewise monstrosity.)

This is happening because Fortran integer division simply drops the decimal part of the number, rather than consistently rounding up or down.  For example, `1.5 / 2` is zero, but `-1.5 / 2` is also zero rather than -1.  If we replace `exponent(x) / 3` with `ceiling(exponent(x) / 3.)` and remove the if-blocks (which now do nothing), then we get a more consistent curve:

new_asx

This new expression does not appear to change answers in your test.

**Comment by Hallberg-NOAA** [comment]
Thank you for all of this helpful feedback, @marshallward. I have incorporated all of this into the updated version of this PR, which is now algorithmically lean and thoroughly commented!

**Comment by Hallberg-NOAA** [comment]
@ashao, thank you for your suggestion with respect to checking the increment size, but in this case it turns out that we have another way to check for convergence and the extra iterations seem pretty cheap, so we can iterate to convergence and then take the better of two near-solutions if the iterations encounter a roundoff-level limit cycle.

File: src/framework/MOM_intrinsic_functions.F90 [hunk 1]
```diff
@@ -25,4 +28,127 @@ function invcosh(x)
...
+  real :: den ! The denominator of an expression for the evolving estimate of the cube root of asx
+              ! in arbitrary units that can grow or shrink with each iteration [C]
+  real :: num_prev ! The numerator of an expression for the previous iteration of the evolving estimate
+              ! of the cube root of asx in arbitrary units that can grow or shrink with each iteration [B D]
+  real :: den_prev ! The denominator of an expression for the previous iteration of the evolving estimate of
+              ! the cube root of asx in arbitrary units that can grow or shrink with each iteration [D]
+  integer :: ex_3 ! One third of the exponent part of x, used to rescale x to get a.
+  integer :: itt
+
+  if ((x >= 0.0) .eqv. (x <= 0.0)) then
```
Comment (marshallward) [inline src/framework/MOM_intrinsic_functions.F90:51]: * Redundant parentheses

* Checking for NaN is not terribly important since the function returns Nan as-is.
  (Zero however does need to be handled separately, since it defies rescaling and solver never converges to it.)

* You could write this as
  ```
  if ((x >= 0.0) .eqv. (x <= 0.0)) return x
  ```
  and avoid the unnecessary indent

File: src/framework/MOM_intrinsic_functions.F90 [hunk 2]
```diff
@@ -25,4 +28,127 @@ function invcosh(x)
...
+    ! Return 0 for an input of 0, or NaN for a NaN input.
+    root = x
+  else
+    ex_3 = ceiling(exponent(x) / 3.)
+    ! Here asx is in the range of 0.125 <= asx < 1.0
+    asx = scale(abs(x), -3*ex_3)
+
+    ! This first estimate is one iteration of Newton's method with a starting guess of 1.  It is
+    ! always an over-estimate of the true solution, but it is a good approximation for asx near 1.
+    num = (2.0 + asx)
```
Comment (marshallward) [inline src/framework/MOM_intrinsic_functions.F90:61]: Redundant parentheses

File: src/framework/MOM_intrinsic_functions.F90 [hunk 3]
```diff
@@ -25,4 +28,127 @@ function invcosh(x)
...
+      ! fewer (or no) real divisions during the iterations before doing a single real division
+      ! at the end, and it is therefore more computationally efficient.
+
+      ! Because successive estimates of the numerator and denominator tend to be the cube of their
+      ! predecessors, the numerator and denominator need to be rescaled by division when they get
+      ! too large or small to avoid overflow or underflow in the convergence test below.  Were
+      ! this being done for 32 bit reals, the values to compare with would be about 1.0e9 and 1.0e-9.
+      if ((den > 1.0e75) .or. (den < 1.0e-75)) then
+        num = num / den ; den = 1.0
+      endif
```
Comment (marshallward) [inline src/framework/MOM_intrinsic_functions.F90:79]: This block could be written as follows:
```
real, parameter :: den_min = 2.**(minexponent(1.) / 4)
real, parameter :: den_max = 2.**(maxexponent(1.) / 4)

! ...

if (den > den_max .or. den < den_min) then
  num = scale(num, -exponent(den))
  den = scale(den, -exponent(den))
endif
````
This would avoid an extra derivative, and also write it in a largely precision-independent form.

I'd also suggest moving it to the end of the loop, since the initial value will never require scaling, and one of the `exit` statements may skip the if-block.
(Depending on the circumstances, it may even be cheaper to do this every iteration, but I'm already overthinking this problem.)

File: src/framework/MOM_intrinsic_functions.F90 [hunk 4]
```diff
@@ -25,4 +28,127 @@ function invcosh(x)
...
+
+      num_prev = num ; den_prev = den
+      num = (2.0 * num_prev**3 + asx * den_prev**3)
```
Comment (marshallward) [inline src/framework/MOM_intrinsic_functions.F90:82]: redundant parentheses

File: src/framework/MOM_intrinsic_functions.F90 [hunk 5]
```diff
@@ -25,4 +28,127 @@ function invcosh(x)
...
+      den = 3.0 * (den_prev * num_prev**2)
```
Comment (marshallward) [inline src/framework/MOM_intrinsic_functions.F90:83]: Just a comment that the order of multiplication appears to affect the final answer here.  But I don't think one is any more accurate than the other.

File: src/framework/MOM_intrinsic_functions.F90 [hunk 6]
```diff
@@ -25,4 +28,127 @@ function invcosh(x)
...
+
+  fail = .false.
+  v = verbose
+  write(stdout,*) '==== MOM_intrinsic_functions: intrinsic_functions_unit_tests ==='
+
+  fail = fail .or. Test_cuberoot(v, 1.2345678901234e9)
+  fail = fail .or. Test_cuberoot(v, -9.8765432109876e-21)
+  fail = fail .or. Test_cuberoot(v, 64.0)
+  fail = fail .or. Test_cuberoot(v, -0.5000000000001)
+  fail = fail .or. Test_cuberoot(v, 0.0)
```
Comment (marshallward) [inline src/framework/MOM_intrinsic_functions.F90:132]: I would add the following tests, since they are important boundary values.
```
fail = fail .or. Test_cuberoot(v, 0.125)
fail = fail .or. Test_cuberoot(v, 1.0)
```

File: src/framework/MOM_intrinsic_functions.F90 [hunk 7]
```diff
@@ -25,4 +28,127 @@ function invcosh(x)
...
+
+      !   If the estimates are increasing, this also indicates convergence, but for a more subtle
+      ! reason.  Because Newton's method converges monotonically from above (at least for infinite
+      ! precision math), the only reason why this estimate could increase is if the iterations
+      ! have converged to a roundoff-level limit cycle around an irrational or otherwise
+      ! unrepresentable solution, with values only changing in the last bit or two.  If so, we
+      ! should stop iterating and accept the one of the current or previous solutions, both of
+      ! which will be within numerical roundoff of the true solution.
+      if (num * den_prev > num_prev * den) then
+        root_asx = num / den
```
Comment (marshallward) [inline src/framework/MOM_intrinsic_functions.F90:99]: This could be written as an `if-else` block, which would avoid a potentially redundant division.

File: src/framework/MOM_intrinsic_functions.F90 [hunk 8]
```diff
@@ -4,9 +4,12 @@ module MOM_intrinsic_functions
 
 ! This file is part of MOM6. See LICENSE.md for the license.
 
+use iso_fortran_env, only : stdout=>output_unit, stderr=>error_unit
```
Comment (marshallward) [inline src/framework/MOM_intrinsic_functions.F90:7]: Spaces around `=>` operator?

File: src/framework/MOM_intrinsic_functions.F90 [hunk 9]
```diff
@@ -25,4 +28,127 @@ function invcosh(x)
...
+      ! this being done for 32 bit reals, the values to compare with would be about 1.0e9 and 1.0e-9.
+      if ((den > 1.0e75) .or. (den < 1.0e-75)) then
+        num = num / den ; den = 1.0
+      endif
+
+      num_prev = num ; den_prev = den
+      num = (2.0 * num_prev**3 + asx * den_prev**3)
+      den = 3.0 * (den_prev * num_prev**2)
+
+      !   If successive estimates of root are identical, this is an exact or converged solution.
```
Comment (marshallward) [inline src/framework/MOM_intrinsic_functions.F90:85]: not exact (how could you know?), just say "this is a converged solution."

**Review by marshallward — CHANGES_REQUESTED** [review]
I believe that I've added my final comments.  It seems to work very well.

I have not checked whether it gives the same answers across compilers and hardware, but we should probably aspire to this at some point.

File: src/framework/MOM_intrinsic_functions.F90 (hunk 3 above)
Comment (Hallberg-NOAA) [reply src/framework/MOM_intrinsic_functions.F90:79]: Thank you for these suggestion; I have now implemented them in the revised PR.

File: src/framework/MOM_intrinsic_functions.F90 (hunk 6 above)
Comment (Hallberg-NOAA) [reply src/framework/MOM_intrinsic_functions.F90:132]: Done.

File: src/framework/MOM_intrinsic_functions.F90 (hunk 8 above)
Comment (Hallberg-NOAA) [reply src/framework/MOM_intrinsic_functions.F90:7]: Corrected.

File: src/framework/MOM_intrinsic_functions.F90 (hunk 9 above)
Comment (Hallberg-NOAA) [reply src/framework/MOM_intrinsic_functions.F90:85]: The comment was changed as suggested.

File: src/framework/MOM_intrinsic_functions.F90 (hunk 7 above)
Comment (Hallberg-NOAA) [reply src/framework/MOM_intrinsic_functions.F90:99]: Thank you for this suggestion; it has been implemented.

File: src/framework/MOM_intrinsic_functions.F90 (hunk 4 above)
Comment (Hallberg-NOAA) [reply src/framework/MOM_intrinsic_functions.F90:82]: Removed.

File: src/framework/MOM_intrinsic_functions.F90 (hunk 1 above)
Comment (Hallberg-NOAA) [reply src/framework/MOM_intrinsic_functions.F90:51]: This is overthinking the problem, and I think that the extra parentheses add readability.  You are welcome to put in a later PR fixing this line if you really feel strongly about it.

File: src/framework/MOM_intrinsic_functions.F90 [hunk 10]
```diff
@@ -25,4 +28,131 @@ function invcosh(x)
...
+  v = verbose
+  write(stdout,*) '==== MOM_intrinsic_functions: intrinsic_functions_unit_tests ==='
+
+  fail = fail .or. Test_cuberoot(v, 1.2345678901234e9)
+  fail = fail .or. Test_cuberoot(v, -9.8765432109876e-21)
+  fail = fail .or. Test_cuberoot(v, 64.0)
+  fail = fail .or. Test_cuberoot(v, -0.5000000000001)
+  fail = fail .or. Test_cuberoot(v, 0.0)
+  fail = fail .or. Test_cuberoot(v, 1.0)
+  fail = fail .or. Test_cuberoot(v, 0.125)
```
Comment (marshallward) [inline src/framework/MOM_intrinsic_functions.F90:136]: Very last suggestion!  This block is untested:
```
if ( abs(den_prev**3*root_asx**3 - den_prev**3*asx) > abs(num_prev**3 - den_prev**3*asx) ) then
  ! The previous iteration was slightly more accurate, so use that for root_asx.
  root_asx = num_prev / den_prev
endif
```
This test should cover it:
```
fail = fail .or. Test_cuberoot(v, 0.965)
```

File: src/framework/MOM_intrinsic_functions.F90 [hunk 11]
```diff
@@ -25,4 +28,131 @@ function invcosh(x)
...
+          ! The previous iteration was slightly more accurate, so use that for root_asx.
+          root_asx = num_prev / den_prev
+        endif
+        exit
+      endif
+
+      ! Because successive estimates of the numerator and denominator tend to be the cube of their
+      ! predecessors, the numerator and denominator need to be rescaled by division when they get
+      ! too large or small to avoid overflow or underflow in the convergence test below.  Were
+      ! this being done for 32 bit reals, the values to compare with would be about 1.0e9 and 1.0e-9.
```
Comment (marshallward) [inline src/framework/MOM_intrinsic_functions.F90:105]: The line about 32-bit reals can probably be removed now.

**Review by marshallward — APPROVED** [review]
Everything looks good!

**Comment by marshallward** [comment]
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/21886 :heavy_check_mark:
